src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            foreach (EnvDTE.CodeElement element in elements)
+            {
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementNamespace)
+                {
+                    if (TryGetCodeElementToRename(oldName, element.Children, out codeElementToRename))
+                    {
+                        return true;
+                    }
+                }
+
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementStruct ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementClass ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementModule ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementInterface ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementEnum)
+                {
+                    string elementName = element.Name;
+                    if (SyntaxFactsService.StringComparer.Equals(oldName, elementName) &&
+                        element is EnvDTE80.CodeElement2 element2)
+                    {
+                        codeElementToRename = element2;
+                        return true;
+                    }
+                }
+            }
+
+            codeElementToRename = null;
+            return false;
+        }
+
+        private async Task<ISymbol> TryGetSymbolToRenameAsync(string oldName,
+                                                              string newFileName,
+                                                              CancellationToken cancellationToken = default)
+        {
+            if (!TryGetDocument(newFileName, out Document newDocument))
+                return null;
+
+            SyntaxNode root = await newDocument.GetSyntaxRootAsync(cancellationToken);
+            if (root is null)
+                return null;
+
+            SemanticModel semanticModel = await newDocument.GetSemanticModelAsync();
+            if (semanticModel is null)
+                return null;
+
+            IEnumerable<SyntaxNode> declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(semanticModel, n, oldName, cancellationToken));
+            SyntaxNode declaration = declarations.FirstOrDefault();
+            if (declaration is null)
+                return null;
+
+            return semanticModel.GetDeclaredSymbol(declaration);
+        }
+
+        private bool  TryGetDocument(string filePath, out Document document)
+        {
+            // NOTE: It is safe to grab the first project we encounter which
+            // contains the given file. Roslyn will handle the case where the
+            // file is included in multiple projects (linked-files or multi-TFM)
+            document = null;
+            DocumentId id = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).FirstOrDefault();
+            if (id is null)
+                return false;
+
+            document = _workspace.CurrentSolution.GetDocument(id);
+            return true;
```

Comment:
NB: since you're doing a blind FirstOrDefault(), you might be getting an document that's not in your project, which could have been asynchronously ripped out of CurrentSolution between this call to GetDocument() and the GetDocumentIdsWithFilePath() earlier since you're not reusing a Solution.